tests/cancun/eip4788_beacon_root/test_beacon_root_precompile.py

```diff
@@ -0,0 +1,250 @@
+    pre: Dict,
+    tx: Transaction,
+    post: Dict,
+):
+    """
+    Tests the beacon root precompile call where the beacon root is equal to the timestamp.
+
+    The expected result is that the precompile call will return the `parent_beacon_block_root`,
+    as all timestamps used are valid.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "call_gas, valid_call",
+    [
+        (BEACON_ROOT_PRECOMPILE_GAS, True),
+        (BEACON_ROOT_PRECOMPILE_GAS + 1, True),
+        (BEACON_ROOT_PRECOMPILE_GAS - 1, False),
+    ],
+)
+@pytest.mark.parametrize(
+    "timestamp",
+    [
+        12,  # twelve
+        2**32,  # arbitrary
+        2**64 - 2,  # near-max
+        2**64 - 1,  # max
+    ],
+)
+@pytest.mark.valid_from("Cancun")
+def test_beacon_root_timestamp_collisions(
+    blockchain_test: BlockchainTestFiller,
+    env: Environment,
+    pre: Dict,
+    tx: Transaction,
+    precompile_call_account: Account,
+    timestamp: int,
+    valid_call: bool,
+):
+    """
+    Tests multiple beacon root precompile calls where the timestamp index is calculated to
+    be equal for each call (i.e colliding). For each parameterized timestamp a list of
+    colliding timestamps are calculated using factors of the `HISTORY_ROOTS_MODULUS`.
+
+    The expected result is that precompile call will return an equal beacon_root
+    for each timestamp used within the call, as the timestamp index used will be the same.
+
+    Here we are predominantly testing that the `timestamp_index` and `root_index` are derived
+    correctly in the evm.
+    """
+    post = {}
+    blocks, colliding_timestamps = [], []
+    timestamp_collisions = 5
+    for i in range(timestamp_collisions):
+        pre[to_address(0x100 + i)] = precompile_call_account
+        colliding_timestamps.append(timestamp + i * HISTORICAL_ROOTS_MODULUS)
+
+    # check timestamp_index function is working as expected
+    timestamp_indexes = [timestamp_index(v) for v in colliding_timestamps]
+    assert len(set(timestamp_indexes)) == 1, "timestamp_index function is not working"
+
+    for i, timestamp in enumerate(colliding_timestamps):
+        blocks.append(
+            Block(
+                txs=[
+                    tx.with_fields(
+                        nonce=i,
+                        to=to_address(0x100),
+                        data=to_hash_bytes(timestamp),
+                    )
+                ],
+                beacon_root=DEFAULT_BEACON_ROOT_HASH,
+            )
+        )
+        post[to_address(0x100 + i)] = Account(
+            storage=expected_storage(
+                beacon_root=DEFAULT_BEACON_ROOT_HASH,
+                timestamp=timestamp,
+                valid_call=valid_call,
+                valid_input=True,
+            )
+        )
+
+    blockchain_test(
+        genesis_environment=env,
+        pre=pre,
+        blocks=blocks,
+        post=post,
+    )
```

Comment:
Add blockchain test where we:
- Overwrite the timestamp/beacon root: `Block2.timestamp == Block1.Timestamp + HISTORICAL_ROOTS_MODULUS`
- Almost overwrite the timestamp/beacon root: `Block2.timestamp == Block1.Timestamp + HISTORICAL_ROOTS_MODULUS + 1` and `Block2.timestamp == Block1.Timestamp + HISTORICAL_ROOTS_MODULUS - 1`
- Set the zero key storage: `Block.timestamp % HISTORICAL_ROOTS_MODULUS == 0`
- Overwrite the zero key storage: `Block2.timestamp == Block1.Timestamp + HISTORICAL_ROOTS_MODULUS && Block1.timestamp % HISTORICAL_ROOTS_MODULUS == 0`